### DISCUSSION TITLE
Catch-all but only for undelivered mail

### DISCUSSION THREAD

**Discussion by MarcS1975** [docker-mailserver/docker-mailserver#3765] [Q&A]
Is it possible to setup a catch-all but **only for the undelivered** mail, so the catchall should not redirect mail for existing users.

**Comment by polarathene** [comment]
It's possible, but it's presently a hassle to configure for.

---

> so the catchall should not redirect mail for existing users.

- `luser_relay = catchall@example.com`
  - The `local` delivery agent is for the DMS hostname (eg: `mail.example.com`), usually system mail / notifications.
  - This should be to a valid mailbox, in DMS use `setup email add` for example, or a remote mail address.
- `@example.com catchall@example.com` at the end of your `postfix-virtual.cf` file.
  - This file is typically managed via our `setup alias ...` commands. You can edit it directly.
  - It must come after any earlier virtual aliases, as the first successful match takes precedence.
  - When `postfix-virtual.cf` is updated (_unless using LDAP at present_), it should be processed and a copy made to `/etc/postfix/virtual`. This matches the Postfix config:
    `virtual_alias_maps = texthash:/etc/postfix/virtual`
  - Alternatively, you can provide your own additional `texthash:/path/to/file` lookup table to the `virtual_alias_maps` setting. This will ensure it's always checked after the first file, allowing you to use `setup alias ...` commands without that sort order concern.
    
    To do this, you can use our `postfix-main.cf` override file support, and include the updated setting. For example if we add `alias-catch-all.cf` as a new file to our DMS config volume:
   `virtual_alias_maps = texthash:/etc/postfix/virtual, texthash:/tmp/docker-mailserver/alias-catch-all.cf`
   Then start your container (_`postfix-main.cf` does not attempt to update automatically, it's only run during container startup_), and your catch-all should be working.

---

**NOTE:**
- The virtual aliases should also apply to `local` before it resolves it's own internal/local-only `/etc/aliases` table.
- Virtual aliases can resolve recipient for outbound mail too IIRC, this is more of an issue when using the similar support for a virtual alias regexp config if the pattern is too relaxed.
- There is a [known problem](https://github.com/docker-mailserver/docker-mailserver/issues/3751#issuecomment-1879323213) (_with [tracking issue](https://github.com/docker-mailserver/docker-mailserver/issues/3022)_).
  - Virtual aliases resolve before virtual mailboxes (_Dovecot managed, these are what DMS accounts are using via `postfix-accounts.cf`_).
  - A catchall will be the fallback when no other virtual alias matches. Now your DMS accounts never receive mail as the catchall was hit before they got a chance. The workaround is to redundantly create an explicit alias for each DMS account.
  - However `setup alias add` no longer supports a virtual alias that overlaps with a DMS account, as that has caused problems in the past for some users (misuse), so this requires manual management similar to the catchall advice/gotcha described for `postfix-virtual.cf`.
